src/setup_payload/AdditionalDataPayloadGenerator.cpp

```diff
@@ -0,0 +1,134 @@
+        mHashInput = nullptr;
+    }
+}
+
+CHIP_ERROR AdditionalDataPayloadGenerator::generateAdditionalDataPayload(PacketBufferHandle & bufferHandle,
+                                                                         bool enableRotatingDeviceId)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    System::PacketBufferTLVWriter writer;
+    TLVWriter innerWriter;
+    char rotatingDeviceIdBuffer[ROTATING_DEVICE_ID_LENGTH * 2 + 1];
+    size_t rotatingDeviceIdBufferSize = 0;
+
+    // Initialize TLVWriter
+    writer.Init(PacketBuffer::New());
+
+    err = writer.OpenContainer(AnonymousTag, kTLVType_Structure, innerWriter);
+    SuccessOrExit(err);
+
+    if (enableRotatingDeviceId)
+    {
+        // Generating Device Rotating Id
+        err = generateRotatingDeviceId(rotatingDeviceIdBuffer, rotatingDeviceIdBufferSize);
+        SuccessOrExit(err);
+
+        // Adding the rotating device id to the TLV data
+        err = innerWriter.PutString(ContextTag(kRotatingDeviceIdTag), rotatingDeviceIdBuffer);
+        SuccessOrExit(err);
+    }
+
+    err = writer.CloseContainer(innerWriter);
+    SuccessOrExit(err);
+
+    err = writer.Finalize(&bufferHandle);
+    SuccessOrExit(err);
+
+exit:
+    return err;
+}
+
+CHIP_ERROR AdditionalDataPayloadGenerator::generateRotatingDeviceId(char * rotatingDeviceIdBuffer,
+                                                                    size_t & rotatingDeviceIdBufferSize)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    char rotationCounterBuf[ROTATING_DEVICE_ID_COUNTER_STR_LENGTH];
+
+    char outputBuffer[ROTATING_DEVICE_ID_LENGTH];
+    uint8_t hashOutputBuffer[kSHA256_Hash_Length];
+    BufferWriter hashInputBufferWriter(reinterpret_cast<unsigned char *>(mHashInput),
+                                       ROTATING_DEVICE_ID_COUNTER_STR_LENGTH + mSerialNumberBufferSize);
+    BufferWriter outputBuferWriter(reinterpret_cast<unsigned char *>(&outputBuffer[0]),
+                                   sizeof(outputBuffer) / sizeof(outputBuffer[0]));
+
+    snprintf(rotationCounterBuf, sizeof(rotationCounterBuf), "%09u", mRotationCounter);
+    hashInputBufferWriter.Put(rotationCounterBuf, ROTATING_DEVICE_ID_COUNTER_STR_LENGTH);
+    hashInputBufferWriter.Put(mSerialNumberBuffer, mSerialNumberBufferSize);
+
+    err = Hash_SHA256(hashInputBufferWriter.Buffer(), ROTATING_DEVICE_ID_COUNTER_STR_LENGTH + mSerialNumberBufferSize,
+                      hashOutputBuffer);
+    SuccessOrExit(err);
+
+    // Computing the Rotating Device Id
+    // RDI = Rotation_Counter + SuffixBytes(HMAC_SHA256(Serial_Number + Rotation_Counter), 16)
+    outputBuferWriter.Put(rotationCounterBuf);
+    outputBuferWriter.Put(
+        reinterpret_cast<const char *>(&hashOutputBuffer[kSHA256_Hash_Length - ROTATING_DEVICE_ID_HASH_SUFFIX_LENGTH]));
+
+    for (size_t i = 0; i < ROTATING_DEVICE_ID_LENGTH; i++)
+    {
+        snprintf(&rotatingDeviceIdBuffer[i * 2], 3, "%02X", outputBuferWriter.Buffer()[i]);
```

Comment:
What ensures that `rotatingDeviceIdBuffer` is big enough for this?

Reply:
good point, i added a check for the size